paddle/framework/data_transform.h

```diff
@@ -0,0 +1,104 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <functional>
+#include <utility>
+#include <vector>
+
+#include "paddle/framework/op_kernel_type.h"
+#include "paddle/framework/tensor.h"
+#include "paddle/platform/device_context.h"
+#include "paddle/platform/macros.h"
+
+namespace paddle {
+namespace framework {
+
+using DataTransformFN =
+    std::function<void(const std::vector<platform::DeviceContext*> ctx,
+                       const Tensor& in, Tensor* out)>;
+using KernelTypePair = std::pair<OpKernelType, OpKernelType>;
+
+struct KernelTypePairHash {
+  size_t operator()(const KernelTypePair& kernel_pair) const {
+    OpKernelType::Hash kernel_type_hasher;
+    size_t left_hasher = kernel_type_hasher(kernel_pair.first) << 1;
+    size_t right_hasher = kernel_type_hasher(kernel_pair.second);
+    std::hash<size_t> hasher;
+    return hasher(left_hasher + right_hasher);
```

Comment:
A little concern on the correctness of hash calculation.

Reply:
ok, I find a better way to do hash

Reply:
thanks, done